### PR TITLE
[DOCS] Add docs for several APIs

### PR DESCRIPTION
**Pull request by reminisce**
@zackchase @madjam @nswamy

### CODE REVIEW

File: python/mxnet/ndarray.py [hunk 1]
```diff
@@ -443,12 +458,26 @@ def _slice(self, start, stop):
         return NDArray(handle=handle, writable=self.writable)
 
     def _at(self, idx):
-        """Returns a sliced view of this array.
+        """Returns a view of the array sliced at idx in the first dim.
```
Comment (piiswrong) [inline python/mxnet/ndarray.py]: explain that this is usually called through x[i]. show this in example.

same for the other apis.

Comment (reminisce) [inline python/mxnet/ndarray.py]: Done.

File: python/mxnet/ndarray.py [hunk 2]
```diff
@@ -403,12 +403,26 @@ def __getitem__(self, key):
 
 
     def _sync_copyfrom(self, source_array):
-        """Peforms a synchronized copy from the array.
+        """Performs a synchronized copy from the source_array to the current array.
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: Place source_array in backticks

File: python/mxnet/ndarray.py [hunk 3]
```diff
@@ -403,12 +403,26 @@ def __getitem__(self, key):
...
+        This is called through x[:] = source_array, where the source_array
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: code snippets should be in double backticks: e.g. "x[:] = source_array"

File: python/mxnet/ndarray.py [hunk 4]
```diff
@@ -403,12 +403,26 @@ def __getitem__(self, key):
...
+        is a numpy.ndarray or array_like object.
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: numpy.ndarray should be in single backticks (clas) 

"array_like" -> "array-like"

File: python/mxnet/ndarray.py [hunk 5]
```diff
@@ -403,12 +403,26 @@ def __getitem__(self, key):
...
+        This function blocks until all the pending read/write operations with respect
+        to the current NDArray are finished and carry out the copy operation to the
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: NDArray in single backticks (in both places)

File: python/mxnet/ndarray.py [hunk 6]
```diff
@@ -426,14 +440,27 @@ def _sync_copyfrom(self, source_array):
             ctypes.c_size_t(source_array.size)))
 
     def _slice(self, start, stop):
-        """Returns a sliced NDArray that shares memory with current one.
+        """Returns a sliced NDArray that shares memory with the current one.
+        This is called through x[start:stop].
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: code in double backticks

File: python/mxnet/ndarray.py [hunk 7]
```diff
@@ -426,14 +440,27 @@ def _sync_copyfrom(self, source_array):
...
-            Starting index of slice.
+            Starting inclusive index of slice in the first dim.
         stop : int
-            Finishing index of slice.
+            Finishing exclusive index of slice in the first dim.
+
+        Returns
+        -------
+            `NDArray` sharing the memory with the current one sliced from
+            start to end in the first dim.
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: you mean "start" to "stop"? start and stop are both parameter names so they should be in single backticks.

File: python/mxnet/ndarray.py (hunk 1 above)
Comment (zackchase) [inline python/mxnet/ndarray.py]: idx is a parameter name so it should be in single backticks.

File: python/mxnet/ndarray.py [hunk 8]
```diff
@@ -443,12 +470,27 @@ def _slice(self, start, stop):
         return NDArray(handle=handle, writable=self.writable)
 
     def _at(self, idx):
-        """Returns a sliced view of this array.
+        """Returns a view of the array sliced at idx in the first dim.
+        This is called through x[idx].
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: this is a code snippet should be in double backticks.

File: python/mxnet/ndarray.py [hunk 9]
```diff
@@ -443,12 +470,27 @@ def _slice(self, start, stop):
...
 
         Parameters
         ----------
         idx : int
-            index of sub array.
+            index for slicing the NDArray in the first dim.
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: NDArray should be in single backticks.

File: python/mxnet/ndarray.py [hunk 10]
```diff
@@ -443,12 +470,27 @@ def _slice(self, start, stop):
...
+
+        Returns
+        -------
+        NDArray
+            `NDArray` sharing the memory with the current one sliced at idx in the first dim.
```
Comment (zackchase) [inline python/mxnet/ndarray.py]: idx is a variable name, single backticks.

File: python/mxnet/test_utils.py [hunk 11]
```diff
@@ -525,7 +537,8 @@ def check_symbolic_forward(sym, location, expected, rtol=1E-4, atol=None,
 
 def check_symbolic_backward(sym, location, out_grads, expected, rtol=1e-5, atol=None,
                             aux_states=None, grad_req='write', ctx=None):
-    """Compare backward call to expected value.
+    """Compares a symbol's backward results with the expected ones.
+    Prints error messages if the backward results are not the same as the expected one.
```
Comment (zackchase) [inline python/mxnet/test_utils.py]: "expected one." -> "expected results."

**Comment by zackchase** [comment]
Hey can we get these changes made and merge? @reminisce

**Comment by reminisce** [comment]
@zackchase Sorry I have been busy with other work. Just finished according to your comments.

**Pull request merged**